### PR TITLE
 Using (processId, threadId) as key to mantain threadpool executor instead of only threadId

### PR DESCRIPTION
**Pull request by gaoyangxiaozhu**
- [x] There's a descriptive title that will make sense to other developers some time from now. 
- [x] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [x] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [x] You have included any necessary tests in the same PR.

This PR attempts to fix the issue [792](https://github.com/dotnet/spark/issues/792) , instead of let multiple `DotnetBackendHandler` instances share one global standalone threadPool object to maintain the relationship between thread and thread id, it always new an new threadPool instance for each `DotnetBackendHandler` instance.

I have already verified and tested end to end in my side.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/spark?pullRequest=793) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/spark?pullRequest=793) before we can accept your contribution.<br/><br/>:x: gaoyangxiaozhu [sign now](https://cla.dotnetfoundation.org/dotnet/spark?pullRequest=793)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/spark?pullRequest=793) it.</sub>

**Comment by gaoyangxiaozhu** [comment]
@suhsteve and @rapoth could you have a look about this PR. Thanks.

**Comment by rapoth** [comment]
Thank you for your contribution @gaoyangxiaozhu! We will take a look next week (some of the folks are on vacation this week).

**Comment by gaoyangxiaozhu** [comment]
hey @suhsteve , so I guess that after another PR [801](https://github.com/dotnet/spark/pull/801) been merged, we can update this PR to address singleton threadPool issue, right?

**Review by imback82 — COMMENTED** [review]
So the main issue here is that the different processes on the client side can share the `ThreadPool` right?  If so, the key of `ThreadPool.executors` can be updated to `processId + threadId` instead of just `threadId` to fix this as well?

The current approach seems fine if we can guarantee that one instance of `DotnetBackendHandler` is assigned to a single client process.

**Comment by gaoyangxiaozhu** [comment]
> So the main issue here is that the different processes on the client side can share the `ThreadPool` right? If so, the key of `ThreadPool.executors` can be updated to `processId + threadId` instead of just `threadId` to fix this as well?
> 
> The current approach seems fine if we can guarantee that one instance of `DotnetBackendHandler` is assigned to a single client process.

@imback82 , thanks Terry, i will update the PR to using `pid+threadId` to represents one thread, will update the PR asap, and let you know.

**Comment by gaoyangxiaozhu** [comment]
@imback82 and @rapoth  I update the PR to use `pid` + `threadId` to represent the thread in DotnetBackend just follow what you suggested, could you help to review the code ? Thanks!

File: src/csharp/Microsoft.Spark/Interop/Ipc/JvmThreadPoolGC.cs [hunk 1]
```diff
@@ -97,16 +98,17 @@ public bool TryAddThread(Thread thread)
...
+        private bool TryDisposeJvmThread(int threadId, int pid)
         {
             if (_activeThreads.TryRemove(threadId, out _))
             {
                 // _activeThreads does not have ownership of the threads on the .NET side. This 
                 // class does not need to call Join() on the .NET Thread. However, this class is
                 // responsible for sending the rmThread command to the JVM to trigger disposal
                 // of the corresponding JVM thread.
-                if ((bool)_jvmBridge.CallStaticJavaMethod("DotnetHandler", "rmThread", threadId))
+                if ((bool)_jvmBridge.CallStaticJavaMethod("DotnetHandler", "rmThread", pid + "_" + threadId))
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Interop/Ipc/JvmThreadPoolGC.cs:111]: Shall we just pass process id separately?

Comment (gaoyangxiaozhu) [reply src/csharp/Microsoft.Spark/Interop/Ipc/JvmThreadPoolGC.cs:111]: hey @imback82 , sure, **sperate process id to pass the dotnetbackend** is also one way to fix the issue.

My understand is that current implement is also ok, right? for fix current issue, seperate process id seems like not bring more benefits while needs more code change?

Or do you have any strong idea to share to use seperate process id? :)

Comment (imback82) [reply src/csharp/Microsoft.Spark/Interop/Ipc/JvmThreadPoolGC.cs:111]: Yea, the current approach is fine. But, having a separate process id seems better for the maintainability reason. I already see three different places (including test) where concatenation logic is introduced. Also, on the server side, if we want to log the info, we need to have a logic to separate process and thread id for a better message. (I also suggest using a tuple as a key (process id, thread id) to make it clear; currently if I look at `ThreadPool` implementation, I am not sure what is inside `threadId: String`.).

I don't mind having a bigger diff as long as we fix it the right way.

Btw, introducing the process id fix the issue right?

Comment (gaoyangxiaozhu) [reply src/csharp/Microsoft.Spark/Interop/Ipc/JvmThreadPoolGC.cs:111]: got it! that make sense, @imback82 , thank you very much for the suggestions, i will update PR to fix it.

Comment (gaoyangxiaozhu) [reply src/csharp/Microsoft.Spark/Interop/Ipc/JvmThreadPoolGC.cs]: hey @imback82 , update the code according your suggestion, please help review when you are available, thank you!

Comment (gaoyangxiaozhu) [reply src/csharp/Microsoft.Spark/Interop/Ipc/JvmThreadPoolGC.cs]: hey, @imback82 , the build check failed  with below error:

Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project microsoft-spark-2-3_2.11: Execution default-compile of goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile failed: Plugin org.apache.maven.plugins:maven-compiler-plugin:3.1 or one of its dependencies could not be resolved: Could not transfer artifact org.apache.maven:maven-core:jar:2.0.9 from/to central (https://repo.maven.apache.org/maven2): Transfer failed for https://repo.maven.apache.org/maven2/org/apache/maven/maven-core/2.0.9/maven-core-2.0.9.jar: Connection reset -> [Help 1]
[ERROR] 

this is the log link:
https://dev.azure.com/dnceng/public/_build/results?buildId=969394&view=logs&j=bd1f0459-543a-55bf-0f8d-4369022127e8&t=a7ce2e4f-2245-52a5-b8ab-5e70fd44d0c0&l=17286

could you help check and solve the issue?

Comment (gaoyangxiaozhu) [reply src/csharp/Microsoft.Spark/Interop/Ipc/JvmThreadPoolGC.cs]: ping @imback82

**Comment by gaoyangxiaozhu** [comment]
@rapoth @imback82, if it is ok, can you help review this PR since I already updated according to @imback82  comments.  Without this PR will block your business scenarios.

File: src/scala/microsoft-spark-2-3/src/main/scala/org/apache/spark/api/dotnet/DotnetBackendHandler.scala [hunk 2]
```diff
@@ -67,9 +68,11 @@ class DotnetBackendHandler(server: DotnetBackend, objectsTracker: JVMObjectTrack
           }
         case "rmThread" =>
           try {
+            assert(serDe.readObjectType(dis) == 'i')
+            val process = serDe.readInt(dis)
```
Comment (suhsteve) [inline src/scala/microsoft-spark-2-3/src/main/scala/org/apache/spark/api/dotnet/DotnetBackendHandler.scala:72]:
```suggestion
            val processId = serDe.readInt(dis)
```

File: src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs [hunk 3]
```diff
@@ -202,13 +203,17 @@ private ISocketWrapper GetConnection()
                 // `StreamingQuery.Stop()` to be called to unblock it. However, the `Stop`
                 // call will never run because DotnetHandler will assign the method call to
                 // run on the same thread that `AwaitTermination` is running on.
+                int processId = Process.GetCurrentProcess().Id;
```
[the remark this answers is not in the document]
Comment (gaoyangxiaozhu) [reply src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs]: yep, update it as readonly memeber

**Comment by gaoyangxiaozhu** [comment]
hey @suhsteve , thank you very much for reviewing, i have updated acccording to your comments, please have a look if you are free

File: src/csharp/Microsoft.Spark/Interop/Ipc/JvmThreadPoolGC.cs [hunk 4]
```diff
@@ -36,11 +38,13 @@ internal class JvmThreadPoolGC : IDisposable
         /// <param name="loggerService">Logger service.</param>
         /// <param name="jvmBridge">The JvmBridge used to call JVM methods.</param>
         /// <param name="threadGCInterval">The interval to GC finished threads.</param>
-        public JvmThreadPoolGC(ILoggerService loggerService, IJvmBridge jvmBridge, TimeSpan threadGCInterval)
+        /// <param name="processId"> The id of current process.</param>
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/Interop/Ipc/JvmThreadPoolGC.cs:41]:
```suggestion
        /// <param name="processId"> The ID of the process.</param>
```

File: src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs [hunk 5]
```diff
@@ -34,6 +35,7 @@ internal sealed class JvmBridge : IJvmBridge
         private const int SocketBufferThreshold = 3;
         private const int ThreadIdForRepl = 1;
 
+        private readonly int processId = Process.GetCurrentProcess().Id;
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs:38]:
```suggestion
        private readonly int _processId = Process.GetCurrentProcess().Id;
```

File: src/scala/microsoft-spark-2-3/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala [hunk 6]
```diff
@@ -39,12 +39,12 @@ object ThreadPool {
 
   /**
    * Try to delete a particular thread.
-   *
+   * @param processId Integer id of process.
```
Comment (suhsteve) [inline src/scala/microsoft-spark-2-3/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala:42]:
```suggestion
   * @param processId Integer id of the process.
```

File: src/scala/microsoft-spark-2-3/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala [hunk 7]
```diff
@@ -62,11 +62,11 @@ object ThreadPool {
 
   /**
    * Get the executor if it exists, otherwise create a new one.
-   *
-   * @param id Integer id of the thread.
+   * @param processId Integer id of process.
```
Comment (suhsteve) [inline src/scala/microsoft-spark-2-3/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala:65]:
```suggestion
   * @param processId Integer id of the process.
```

File: src/scala/microsoft-spark-2-4/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala [hunk 8]
```diff
@@ -39,12 +39,12 @@ object ThreadPool {
 
   /**
    * Try to delete a particular thread.
-   *
+   * @param processId Integer id of process.
```
Comment (suhsteve) [inline src/scala/microsoft-spark-2-4/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala:42]:
```suggestion
   * @param processId Integer id of the process.
```

File: src/scala/microsoft-spark-2-4/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala [hunk 9]
```diff
@@ -62,11 +62,11 @@ object ThreadPool {
 
   /**
    * Get the executor if it exists, otherwise create a new one.
-   *
-   * @param id Integer id of the thread.
-   * @return The new or existing executor with the given id.
+   * @param processId Integer id of process.
```
Comment (suhsteve) [inline src/scala/microsoft-spark-2-4/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala:65]:
```suggestion
   * @param processId Integer id of the process.
```

File: src/scala/microsoft-spark-3-0/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala [hunk 10]
```diff
@@ -62,11 +62,11 @@ object ThreadPool {
 
   /**
    * Get the executor if it exists, otherwise create a new one.
-   *
-   * @param id Integer id of the thread.
+   * @param processId Integer id of process.
```
Comment (suhsteve) [inline src/scala/microsoft-spark-3-0/src/main/scala/org/apache/spark/api/dotnet/ThreadPool.scala:65]:
```suggestion
   * @param processId Integer id of the process.
```

**Review by suhsteve — DISMISSED** [review]
A few minor nits, but LGTM.  Thanks @gaoyangxiaozhu

**Comment by gaoyangxiaozhu** [comment]
updated! thanks @suhsteve, still need you help approve, since the update reset your vote.

and @imback82 , could you help review to see if there any other issue for the PR?

File: src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs [hunk 11]
```diff
@@ -205,10 +207,13 @@ private ISocketWrapper GetConnection()
                 Thread thread = _isRunningRepl ? null : Thread.CurrentThread;
                 MemoryStream payloadMemoryStream = s_payloadMemoryStream ??= new MemoryStream();
                 payloadMemoryStream.Position = 0;
+
+                int threadId = thread == null ? ThreadIdForRepl : thread.ManagedThreadId;
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs:211]: nit: let's move this after 207:
```csharp
Thread thread = _isRunningRepl ? null : Thread.CurrentThread;
int threadId = thread == null ? ThreadIdForRepl : thread.ManagedThreadId;
```

**Review by imback82 — COMMENTED** [review]
LGTM (one nit), thanks @gaoyangxiaozhu!

File: src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs (hunk 11 above)
Comment (gaoyangxiaozhu) [reply src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs]: he y

**Review by imback82 — APPROVED** [review]

**Review by suhsteve — APPROVED** [review]

**Pull request merged**